### PR TITLE
[Menu items REST API] Perform validation and sanitization separately from saving

### PR DESCRIPTION
**Pull request by adamziel**
This PR removes the `prepare_item_for_database` method in favor of more specific validation and sanitization callbacks, with the ultimate goal being supporting batch requests. Let's discuss in comments!

Test plan:
1. Apply https://github.com/WordPress/wordpress-develop/pull/311 to your local WordPress.
1. Manually add support for `sanitize_callback` (see comments on the PR above).
1. Run phpunit tests, confirm they pass.

### CODE REVIEW

File: lib/class-wp-rest-menu-items-controller.php [hunk 1]
```diff
@@ -745,13 +667,33 @@ public function get_item_schema() {
...
-				'sanitize_callback' => null,
-				// Note: validation implemented in self::prepare_item_for_database().
-				'validate_callback' => null,
+				'validate_callback' => static function ( $value, $request ) {
+					if ( 'custom' === $request['type'] && '' === $value ) {
+						return new WP_Error(
+							'rest_title_required',
+							__( 'Title required if menu item of type custom.', 'gutenberg' )
+						);
+					}
```
Comment (adamziel) [inline lib/class-wp-rest-menu-items-controller.php:679]: This code will never be executed if the title is missing. I considered the following:
1. Move it over to `type` validate_callback - doesn't seem right, this error is specific to `title`.
2. Set a default value of `""` to ensure this callback is fired on every request. Far from perfection, let's imagine the title is set in the database but is missing from the request, it would override the stored title with an empty string. 
3. Make this field universally required 
4. Add a concept of "conditionally required" parameter to the REST API.
5. Add a controller-wide `validate_callback` and `sanitize_callback` to deal with cases that don't neatly fit into property-specific ones.

**Comment by adamziel** [comment]
This is not ready for a review yet and needs to be rebased on top of https://github.com/WordPress/wordpress-develop/pull/311 first - validating without using a controller-level `validate_callback` is really far from perfect.

**Comment by adamziel** [comment]
This is now ready for a discussion. I moved field-related validation and sanitization to the schema. This left me with some code that took both the user input and the database data, and transformed it into a saveable `$prepared_nav_item`. This sounds like a sanitization to me, so I moved it to a `sanitize_callback` (it would be great to add support for it to https://github.com/WordPress/wordpress-develop/pull/311).

Code structure wise, I don't love how it creates a synthetic sanitized request property `$request['prepared_nav_item'] = $prepared_nav_item;`. Another solution would be producing `$prepared_nav_item` using `get_nav_menu_item` and `$request` once during the sanitization, and then for the second time in the actual handler method - I don't like this solution either. Any thoughts appreciated @TimothyBJacobs

**Comment by adamziel** [comment]
I also think menu position needs to be handled differently to work with batch processing. It would neatly fit a follow-up PR with some additional test cases.

**Comment by TimothyBJacobs** [comment]
This is looking good overall, I'm going to hold off on detailed feedback for a bit while focusing on Beta 1.

File: lib/class-wp-rest-menu-items-controller.php [hunk 2]
```diff
@@ -368,64 +478,51 @@ protected function prepare_item_for_database( $request ) {
 				$check = rest_validate_value_from_schema( $request[ $api_request ], $schema['properties'][ $api_request ] );
```
Comment (TimothyBJacobs) [inline lib/class-wp-rest-menu-items-controller.php:478]: Let's remove this double schema validation while we're here. The request data has already been validated and sanitized according to the schema.

File: lib/class-wp-rest-menu-items-controller.php [hunk 3]
```diff
@@ -745,13 +789,33 @@ public function get_item_schema() {
...
+					}
+
+					$title = '';
+					if ( is_string( $value ) ) {
+						$title = $value;
+					} elseif ( ! empty( $value['raw'] ) ) {
+						$title = $value['raw'];
+					}
+
+					$title = wp_unslash( apply_filters( 'title_save_pre', wp_slash( $title ) ) );
```
Comment (TimothyBJacobs) [inline lib/class-wp-rest-menu-items-controller.php:815]: Fairly certain this should be removed: #25095.

File: lib/class-wp-rest-menu-items-controller.php [hunk 4]
```diff
@@ -813,7 +891,18 @@ public function get_item_schema() {
 			'type'        => 'string',
 			'context'     => array( 'view', 'edit', 'embed' ),
 			'arg_options' => array(
-				'sanitize_callback' => 'sanitize_text_field',
+				'sanitize_callback' => static function ( $value ) {
+					$sanitized = sanitize_text_field( $value );
+					/** This filter is documented in wp-includes/post.php */
+					$sanitized = wp_unslash(
```
Comment (TimothyBJacobs) [inline lib/class-wp-rest-menu-items-controller.php:897]: Same with this, #25095.

File: lib/class-wp-rest-menu-items-controller.php [hunk 5]
```diff
@@ -836,7 +928,18 @@ public function get_item_schema() {
 			'type'        => 'string',
 			'context'     => array( 'view', 'edit', 'embed' ),
 			'arg_options' => array(
-				'sanitize_callback' => 'sanitize_text_field',
+				'sanitize_callback' => static function ( $value, $request ) {
+					$sanitized = sanitize_text_field( $value );
+					/** This filter is documented in wp-includes/post.php */
+					$sanitized = wp_unslash(
```
Comment (TimothyBJacobs) [inline lib/class-wp-rest-menu-items-controller.php:934]: And this #25095.

File: lib/class-wp-rest-menu-items-controller.php [hunk 6]
```diff
@@ -813,7 +891,18 @@ public function get_item_schema() {
 			'type'        => 'string',
 			'context'     => array( 'view', 'edit', 'embed' ),
 			'arg_options' => array(
-				'sanitize_callback' => 'sanitize_text_field',
+				'sanitize_callback' => static function ( $value ) {
+					$sanitized = sanitize_text_field( $value );
```
Comment (TimothyBJacobs) [inline lib/class-wp-rest-menu-items-controller.php:895]: I'm not sure this function should be called, it was the previous `sanitize_callback`, but AFAICT the title attr is not sanitized with this in Core.

File: lib/class-wp-rest-menu-items-controller.php [hunk 7]
```diff
@@ -883,6 +1014,18 @@ public function get_item_schema() {
 			'type'        => 'string',
 			'format'      => 'uri',
 			'context'     => array( 'view', 'edit', 'embed' ),
+			'arg_options' => array(
+				'validate_callback' => static function ( $value ) {
+					$validated = esc_url_raw( $value );
+					if ( '' === $validated ) {
+						// Fail sanitization if URL is invalid.
+						return new WP_Error( 'invalid_url', __( 'Invalid URL.', 'gutenberg' ), array( 'status' => 400 ) );
+					}
```
Comment (TimothyBJacobs) [inline lib/class-wp-rest-menu-items-controller.php:1023]:
```suggestion
					}
					
					return true;
```

File: lib/class-wp-rest-menu-items-controller.php [hunk 8]
```diff
@@ -474,66 +571,13 @@ protected function prepare_item_for_database( $request ) {
...
-		 * Filters a post before it is inserted via the REST API.
-		 *
-		 * The dynamic portion of the hook name, `$this->post_type`, refers to the post type slug.
-		 *
-		 * @param stdClass        $prepared_post An object representing a single post prepared
-		 *                                       for inserting or updating the database.
-		 * @param WP_REST_Request $request       Request object.
-		 */
-		return apply_filters( "rest_pre_insert_{$this->post_type}", $prepared_nav_item, $request );
+		$request['prepared_nav_item'] = $prepared_nav_item;
```
Comment (TimothyBJacobs) [inline lib/class-wp-rest-menu-items-controller.php:574]: I'm really not a fan of us passing data thru the request object like this. What does it look like to stick with a `prepare_item_for_database` method that does this with already validated data.

**Review by TimothyBJacobs — CHANGES_REQUESTED** [review]
Thanks again for starting work on this! I've left some feedback.

Now that we have #25096, we could refactor this to be based off that branch for the moment. We will need to do a version check to make sure the user is running at least [r48945](https://core.trac.wordpress.org/changeset/48945) which adds support for a route-level `validate_callback` and if the user isn't, manually calling our `validate_callback` in the route callback.

What would this look like if we kept `prepare_item_for_database` and made the `validate_callback` only do validation? I understand there will be some duplicate work, but how much of an impact would that actually have?

Another thing we need to keep in mind, is that the REST API applies validation before doing permission checks. So we're currently leaking some data at the moment. For the moment, I think we can manually call our permission callbacks before doing the more sensitive validation. But I'm thinking thru what a Core solution might look like.

File: lib/class-wp-rest-menu-items-controller.php [hunk 9]
```diff
@@ -894,7 +1037,10 @@ public function get_item_schema() {
 			'context'     => array( 'view', 'edit', 'embed' ),
 			'arg_options' => array(
 				'sanitize_callback' => function ( $value ) {
-					return array_map( 'sanitize_html_class', wp_parse_list( $value ) );
+					$sanitized = $value;
+					$sanitized = array_map( 'sanitize_html_class', wp_parse_list( $sanitized ) );
+
+					return array_map( 'sanitize_html_class', $sanitized );
```
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:1043]: Why does this need change?

File: lib/class-wp-rest-menu-items-controller.php [hunk 10]
```diff
@@ -883,6 +1014,18 @@ public function get_item_schema() {
...
+				'validate_callback' => static function ( $value ) {
+					$validated = esc_url_raw( $value );
+					if ( '' === $validated ) {
+						// Fail sanitization if URL is invalid.
+						return new WP_Error( 'invalid_url', __( 'Invalid URL.', 'gutenberg' ), array( 'status' => 400 ) );
+					}
+				},
+				'sanitize_callback' => static function ( $value ) {
+					return esc_url_raw( $value );
+				},
```
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:1027]:
```suggestion
				'sanitize_callback' => 'esc_url_raw',
```

File: lib/class-wp-rest-menu-items-controller.php [hunk 11]
```diff
@@ -923,6 +1069,19 @@ public function get_item_schema() {
...
+						// Check if nav menu is valid.
+						if ( ! is_nav_menu( absint( $value ) ) ) {
+							return new WP_Error( 'invalid_menu_id', __( 'Invalid menu ID.', 'gutenberg' ), array( 'status' => 400 ) );
+						}
+
+						return true;
+					},
+					'sanitize_callback' => static function ( $value ) {
+						return absint( $value );
+					},
```
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:1083]:
```suggestion
					'sanitize_callback' => 'absint'
```

File: lib/class-wp-rest-menu-items-controller.php [hunk 12]
```diff
@@ -854,11 +957,36 @@ public function get_item_schema() {
...
+						if ( empty( $original ) ) {
+							return new WP_Error( 'rest_post_invalid_id', __( 'Invalid post ID.', 'gutenberg' ), array( 'status' => 400 ) );
+						}
+					}
+
+					return true;
+				},
+				'sanitize_callback' => static function ( $value ) {
+					return intval( $value );
+				},
```
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:988]:
```suggestion
				'sanitize_callback' => 'intval',
```

File: lib/class-wp-rest-menu-items-controller.php [hunk 13]
```diff
@@ -790,6 +854,11 @@ public function get_item_schema() {
 			'enum'        => array( 'taxonomy', 'post_type', 'post_type_archive', 'custom' ),
 			'context'     => array( 'view', 'edit', 'embed' ),
 			'default'     => 'custom',
+			'arg_options' => array(
+				'sanitize_callback' => function ( $value ) {
+					return sanitize_key( $value );
+				},
```
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:860]:
```suggestion
				'sanitize_callback' => 'sanitize_key'
```

File: lib/class-wp-rest-menu-items-controller.php [hunk 14]
```diff
@@ -745,13 +789,33 @@ public function get_item_schema() {
 
 		$schema['properties']['title'] = array(
 			'description' => __( 'The title for the object.', 'gutenberg' ),
-			'type'        => 'object',
+			'type'        => array( 'string', 'object' ),
```
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:792]: Can you explain this change?

File: lib/class-wp-rest-menu-items-controller.php [hunk 15]
```diff
@@ -836,7 +928,18 @@ public function get_item_schema() {
...
+					/** This filter is documented in wp-includes/post.php */
+					$sanitized = wp_unslash(
+						apply_filters(
+							'content_save_pre',
+							wp_slash( $sanitized )
+						)
+					);
+
+					return $sanitized;
+				},
```
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:942]: This code has been repeated. Let's just make it a method so we can we use it.

File: lib/class-wp-rest-menu-items-controller.php [hunk 16]
```diff
@@ -826,7 +915,10 @@ public function get_item_schema() {
 			'context'     => array( 'view', 'edit', 'embed' ),
 			'arg_options' => array(
 				'sanitize_callback' => function ( $value ) {
-					return array_map( 'sanitize_html_class', wp_parse_list( $value ) );
+					$sanitized = $value;
+					$sanitized = array_map( 'sanitize_html_class', wp_parse_list( $sanitized ) );
+
+					return array_map( 'sanitize_html_class', $sanitized );
 				},
```
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:922]: Repeated code again.

File: lib/class-wp-rest-menu-items-controller.php (hunk 12 above)
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:988]: Why is this needed?

File: lib/class-wp-rest-menu-items-controller.php (hunk 10 above)
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:1027]: Why is this required, shouldn't format: uri cover this?

File: lib/class-wp-rest-menu-items-controller.php [hunk 17]
```diff
@@ -923,6 +1069,19 @@ public function get_item_schema() {
...
+				$schema['properties'][ $base ]['arg_options'] = array(
+					'validate_callback' => static function ( $value ) {
+						// Check if nav menu is valid.
+						if ( ! is_nav_menu( absint( $value ) ) ) {
+							return new WP_Error( 'invalid_menu_id', __( 'Invalid menu ID.', 'gutenberg' ), array( 'status' => 400 ) );
+						}
+
+						return true;
+					},
+					'sanitize_callback' => static function ( $value ) {
```
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:1081]: Is this required?

File: lib/class-wp-rest-menu-items-controller.php [hunk 18]
```diff
@@ -186,15 +281,25 @@ public function update_item( $request ) {
 			return $valid_check;
 		}
 
-		$prepared_nav_item = $this->prepare_item_for_database( $request );
+		$prepared_nav_item = $request['prepared_nav_item'];
```
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:284]:
```suggestion
		$prepared_nav_item = $this->sanitize( $request );
```

File: lib/class-wp-rest-menu-items-controller.php [hunk 19]
```diff
@@ -97,7 +181,18 @@ public function create_item( $request ) {
 			return new WP_Error( 'rest_post_exists', __( 'Cannot create existing post.', 'gutenberg' ), array( 'status' => 400 ) );
 		}
 
-		$prepared_nav_item = $this->prepare_item_for_database( $request );
+		$prepared_nav_item = $request['prepared_nav_item'];
```
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:184]:
```suggestion
		$prepared_nav_item = $this->sanitize( $request );
```
This is a bad pattern to put an object into state like this. Just run the method again.

File: lib/class-wp-rest-menu-items-controller.php (hunk 8 above)
Comment (spacedmonkey) [reply lib/class-wp-rest-menu-items-controller.php:574]: Could not agree more just put the return and filter back. This change is completely unnessary

Comment (spacedmonkey) [reply lib/class-wp-rest-menu-items-controller.php:574]: This line seems to completely overwrite the existing request. Please remove as it breaks things.

File: lib/class-wp-rest-menu-items-controller.php [hunk 20]
```diff
@@ -368,64 +478,51 @@ protected function prepare_item_for_database( $request ) {
...
-		}
-
-		$taxonomy = get_taxonomy( 'nav_menu' );
-		$base     = ! empty( $taxonomy->rest_base ) ? $taxonomy->rest_base : $taxonomy->name;
-		// If menus submitted, cast to int.
-		if ( isset( $request[ $base ] ) && ! empty( $request[ $base ] ) ) {
-			$prepared_nav_item['menu-id'] = absint( $request[ $base ] );
-		}
-
-		// Nav menu title.
```
Comment (spacedmonkey) [inline lib/class-wp-rest-menu-items-controller.php:384]: Title no longer works without this code. Why was this removed?

**Comment by adamziel** [comment]
Closing this stale PR, we may need to find another way forward.
